addons/binding/org.openhab.binding.paradoxalarm/src/main/java/org/openhab/binding/paradoxalarm/internal/communication/GenericCommunicator.java

```diff
@@ -0,0 +1,331 @@
+import org.openhab.binding.paradoxalarm.internal.communication.messages.IpMessagesConstants;
+import org.openhab.binding.paradoxalarm.internal.communication.messages.ParadoxIPPacket;
+import org.openhab.binding.paradoxalarm.internal.util.ParadoxUtil;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * The {@link GenericCommunicator} Used for the common communication logic for all types of panels. Future use to
+ * autodetect panels and provide information to factory to create the proper type of commumnicator.
+ *
+ * @author Konstantin_Polihronov - Initial contribution
+ */
+public class GenericCommunicator implements IParadoxGenericCommunicator {
+
+    private static final int SOCKET_TIMEOUT = 4000;
+
+    private static final byte[] LOGOUT_MESAGE_BYTES = new byte[] { 0x00, 0x07, 0x05, 0x00, 0x00, 0x00, 0x00 };
+
+    protected final Logger logger = LoggerFactory.getLogger(GenericCommunicator.class);
+
+    protected Socket socket;
+    protected DataOutputStream tx;
+    protected DataInputStream rx;
+    protected final byte[] pcPasswordBytes;
+    protected byte[] panelInfoBytes;
+    protected boolean isOnline;
+
+    private String ipAddress;
+    private int tcpPort;
+    private String password;
+
+    public GenericCommunicator(String ipAddress, int tcpPort, String ip150Password, String pcPassword)
+            throws UnknownHostException, IOException, InterruptedException {
+        this.ipAddress = ipAddress;
+        this.tcpPort = tcpPort;
+        this.password = ip150Password;
+
+        initializeSocket();
+
+        this.pcPasswordBytes = ParadoxUtil.stringToBCD(pcPassword);
+        loginSequence();
+    }
+
+    private void initializeSocket() throws UnknownHostException, IOException {
+        if (socket == null || socket.isClosed()) {
+            socket = new Socket(ipAddress, tcpPort);
+            socket.setSoTimeout(SOCKET_TIMEOUT);
+            tx = new DataOutputStream(socket.getOutputStream());
+            rx = new DataInputStream(socket.getInputStream());
+        }
+    }
+
+    @Override
+    public synchronized void close() {
+        logger.info("Stopping communication to Paradox system");
+        try {
+            tx.close();
+            rx.close();
+            socket.close();
+            // This is very ugly but Paradox supports only one connection at a time and if not closed properly when
+            // handler gets destroyed/recreated before the full socket closure, the new handler cannot establish proper
+            // communication.
+            logger.info("Waiting the socket to close...");
+            Thread.sleep(1000);
+
+        } catch (InterruptedException e) {
+            logger.warn(
+                    "Unable to sleep thread during socket close phase. Could lead to issues if reconnect occurs. {}",
+                    e);
+        } catch (IOException e) {
+            logger.warn("IO exception during socket/stream close operation. {}", e);
```

Comment:
Only put exception as last param, but don't refer to it in the message string. This is the way to do it for exception logging.
```suggestion
            logger.warn("IO exception during socket/stream close operation. ", e);
```